File: Calto-GG/Goldman Sachs/Day 2/Split_Array_into_Consecutive_Subsequences.py
from collections import Counter

# A bit tricky to proceed with.. initialize the counter to get the frequency 
# of the numbers occuring and dont need to sort here since they are already sorted
# in Counter cause of input is already sorted next we play with pointers inside the 
# dictionay of counter. Time Complexity is O(set(N)^2)
# but its memory efficient no extra stuff required besides the counter dict. 

class Solution:
    def isPossible(self, nums: list[int]) -> bool:
        counter = Counter(nums)
        for i in counter.keys():
            while counter[i] > 0:
                last = 0
                j = i
                k = 0
                while counter[j] >= last:
                    last = counter[j]
                    counter[j] -= 1
                    j += 1
                    k += 1
                if k < 3:
                    return False
        return True

File: Sudesh_R/Goldman Sachs/Day7/split-array-into-consecutive-subsequences.cpp
class Solution {
public:
    bool isPossible(vector<int>& nums) {
    
        map<int,int>m , m1;
        for(auto it : nums){m[it]++;}
        for(int i = 0;i< nums.size() ; i++)
        {
            if( m[nums[i]]==0){continue;}
            m[nums[i]]--;
            if(m1[nums[i]-1] >0)
            {
                m1[nums[i]-1]--;
                m1[nums[i]]++;
            }
            else if(m[nums[i ]+1]  >0 && m[nums[i]+2] >0)
            {
                m[nums[i]+1]--;
                m[nums[i]+2]--;

                m1[nums[i] + 2]++;
            }
            else{return false;}
        }
return true;


    }
};
